### ISSUE TITLE
Failed to Initialize Gradle Language Server

### ISSUE THREAD

**Issue by Drakmyth** [badsyntax/vscode-gradle#1028]
Extension Name: vscode-gradle
Extension Version: v3.8.0
OS Version: Windows 10.0.19043 Build 19043
VSCode version: 1.60.2

**Describe the bug**
Upon opening a Gradle project, the Gradle Language Server throws errors and fails to start properly. The status bar indicates it's stuck in "Initializing Gradle Language Server". Upon opening a *.java file or the Tasks panel, the Language Server tries again to start up and succeeds.

**To Reproduce**
Open any folder with a build.gradle at the root (including this repository itself).

**Expected behavior**
No error should be thrown. Status bar should not be stuck with "Initializing Gradle Language Server" spinner.

**Output from "Gradle Tasks"**
```
[debug] [gradle-server] Starting server
[debug] [gradle-server] Gradle Server cmd: c:\Users\shaun_000\.vscode\extensions\vscjava.vscode-gradle-3.8.0\lib\gradle-server.bat 49258
[info] [gradle-server] Server started, listening on 49258
[info] Gradle client connected to server
```

**Output from "Gradle Language Server"**
```
Oct 05, 2021 9:53:28 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SEVERE: Internal error: java.lang.reflect.InvocationTargetException
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	... 11 more
Caused by: java.lang.NullPointerException: Cannot read the array length because "<local3>" is null
	at com.microsoft.gradle.resolver.GradleLibraryResolver.searchInFolder(GradleLibraryResolver.java:161)
	at com.microsoft.gradle.resolver.GradleLibraryResolver.findCoreAPIWithDist(GradleLibraryResolver.java:152)
	at com.microsoft.gradle.resolver.GradleLibraryResolver.findCoreAPIWithWrapper(GradleLibraryResolver.java:140)
	at com.microsoft.gradle.resolver.GradleLibraryResolver.resolveGradleAPI(GradleLibraryResolver.java:89)
	at com.microsoft.gradle.GradleServices.applySetting(GradleServices.java:160)
	at com.microsoft.gradle.GradleLanguageServer.initialize(GradleLanguageServer.java:90)
	... 16 more

[Error - 9:53:29 PM] Server initialization failed.
  Message: Internal error.
  Code: -32603 
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	... 11 more
Caused by: java.lang.NullPointerException: Cannot read the array length because "<local3>" is null
	at com.microsoft.gradle.resolver.GradleLibraryResolver.searchInFolder(GradleLibraryResolver.java:161)
	at com.microsoft.gradle.resolver.GradleLibraryResolver.findCoreAPIWithDist(GradleLibraryResolver.java:152)
	at com.microsoft.gradle.resolver.GradleLibraryResolver.findCoreAPIWithWrapper(GradleLibraryResolver.java:140)
	at com.microsoft.gradle.resolver.GradleLibraryResolver.resolveGradleAPI(GradleLibraryResolver.java:89)
	at com.microsoft.gradle.GradleServices.applySetting(GradleServices.java:160)
	at com.microsoft.gradle.GradleLanguageServer.initialize(GradleLanguageServer.java:90)
	... 16 more

[Error - 9:53:29 PM] Starting client failed
  Message: Internal error.
  Code: -32603 
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	... 11 more
Caused by: java.lang.NullPointerException: Cannot read the array length because "<local3>" is null
	at com.microsoft.gradle.resolver.GradleLibraryResolver.searchInFolder(GradleLibraryResolver.java:161)
	at com.microsoft.gradle.resolver.GradleLibraryResolver.findCoreAPIWithDist(GradleLibraryResolver.java:152)
	at com.microsoft.gradle.resolver.GradleLibraryResolver.findCoreAPIWithWrapper(GradleLibraryResolver.java:140)
	at com.microsoft.gradle.resolver.GradleLibraryResolver.resolveGradleAPI(GradleLibraryResolver.java:89)
	at com.microsoft.gradle.GradleServices.applySetting(GradleServices.java:160)
	at com.microsoft.gradle.GradleLanguageServer.initialize(GradleLanguageServer.java:90)
	... 16 more
```

**Does the bug still exist if you disable all other extensions?**
Yes

**Additional context**
The below issue is unrelated, but was including in case that wasn't the case. Thanks to @CsCherrYY for confirming it's unrelated to the above.
~After waiting for a couple minutes after opening the `vscode-gradle` repository folder, the following errors popped up. I haven't seen these before so I suspect they may be related to the vscode-gradle repository itself rather than this issue (which applies to all gradle projects) but including them just in case:~
<details>
<summary>Images of the unrelated issue</summary>
![image](https://user-images.githubusercontent.com/7451023/136144151-8c1112c5-e64c-4f35-9151-c08e52633a05.png)
</details>

**Comment by yagnasrinath** [comment]
Any workaround for this issue?

**Comment by CsCherrYY** [comment]
You will see this pop up since your Gradle build is not meet your Java runtime requirement, see https://docs.gradle.org/current/userguide/compatibility.html for details.

As it suggested, you can specific `java.import.gradle.java.home` to a lower JDK version (e.g. Java 11) and reload window to see if this works.

**Comment by Drakmyth** [comment]
@yagnasrinath Not that I've found so far. It's definitely something with my local environment configuration, as I have a second machine that does not experience this error, but I've been unable to determine what. Since the server does start correctly on the second attempt, I haven't found this to be a blocker, but it's clear something isn't functioning as expected.

@CsCherrYY That explains the pop ups, but confirms they are unrelated to the main issue. I'll scratch that out.

**Comment by yagnasrinath** [comment]
Yeah, it happens on one of my ubuntu machines with java 8. So, I do not think java version is an issue in my case. Nevertheless I tried to set the path and it didn't seem to help resolve the issue.

**Comment by CsCherrYY** [comment]
@Drakmyth Thanks for the clarifying, I think this issue is related to the library resolver in the language server, we use the resolver to parse the gradle distribution you are using to provide some auto completion candidates for you. Are you using a gradle wrapper in your project?

@yagnasrinath Could you please provide the log from **Gradle Language Server** channel? I'd like to check if it comes from the root cause as above.

**Comment by Drakmyth** [comment]
Yes, I've only ever tried opening projects that have a gradle wrapper. If I delete the `gradlew`, `gradlew.bat`, `gradle/wrapper/gradle-wrapper.jar`, and `gradle/wrapper/gradle-wrapper.properties` files, the Language Server no longer throws errors upon opening the project.

Putting those files back causes the error to be thrown again.

**Comment by yagnasrinath** [comment]
The only log I see is:
```
Exception in thread "main" java.lang.NoSuchMethodError: java.util.List.of(Ljava/lang/Object;Ljava/lang/Object;)Ljava/util/List;
	at com.microsoft.gradle.GradleServices.<clinit>(GradleServices.java:86)
	at com.microsoft.gradle.GradleLanguageServer.<init>(GradleLanguageServer.java:76)
	at com.microsoft.gradle.GradleLanguageServer.main(GradleLanguageServer.java:54)
[Info  - 8:15:03 PM] Connection to server got closed. Server will restart.
Exception in thread "main" java.lang.NoSuchMethodError: java.util.List.of(Ljava/lang/Object;Ljava/lang/Object;)Ljava/util/List;
	at com.microsoft.gradle.GradleServices.<clinit>(GradleServices.java:86)
	at com.microsoft.gradle.GradleLanguageServer.<init>(GradleLanguageServer.java:76)
	at com.microsoft.gradle.GradleLanguageServer.main(GradleLanguageServer.java:54)
[Info  - 8:15:03 PM] Connection to server got closed. Server will restart.
Exception in thread "main" java.lang.NoSuchMethodError: java.util.List.of(Ljava/lang/Object;Ljava/lang/Object;)Ljava/util/List;
	at com.microsoft.gradle.GradleServices.<clinit>(GradleServices.java:86)
	at com.microsoft.gradle.GradleLanguageServer.<init>(GradleLanguageServer.java:76)
	at com.microsoft.gradle.GradleLanguageServer.main(GradleLanguageServer.java:54)
[Info  - 8:15:03 PM] Connection to server got closed. Server will restart.
Exception in thread "main" java.lang.NoSuchMethodError: java.util.List.of(Ljava/lang/Object;Ljava/lang/Object;)Ljava/util/List;
	at com.microsoft.gradle.GradleServices.<clinit>(GradleServices.java:86)
	at com.microsoft.gradle.GradleLanguageServer.<init>(GradleLanguageServer.java:76)
	at com.microsoft.gradle.GradleLanguageServer.main(GradleLanguageServer.java:54)
[Info  - 8:15:04 PM] Connection to server got closed. Server will restart.
Exception in thread "main" java.lang.NoSuchMethodError: java.util.List.of(Ljava/lang/Object;Ljava/lang/Object;)Ljava/util/List;
	at com.microsoft.gradle.GradleServices.<clinit>(GradleServices.java:86)
	at com.microsoft.gradle.GradleLanguageServer.<init>(GradleLanguageServer.java:76)
	at com.microsoft.gradle.GradleLanguageServer.main(GradleLanguageServer.java:54)
[Error - 8:15:04 PM] Connection to server got closed. Server will not be restarted.
```
I am not sure how to enable debug logging. Let me know if there is a way to enable debug logging. I also have gradle wrapper in my project. And gradle language server in my case never successfully initializes though.

**Comment by CsCherrYY** [comment]
@yagnasrinath Your issue might be related to https://github.com/microsoft/vscode-gradle/issues/1024#issuecomment-930692759 and there is a fix in #1029.

**Comment by yagnasrinath** [comment]
@CsCherrYY you are right. My issue is related to java 8. I got it fixed by upgrading to java 11. Thanks for pointing me to the fix.

**Comment by CsCherrYY** [comment]
@yagnasrinath Sorry about the inconvenience. The fix will be available in `3.8.1` and you can use Java 8 back since it has been released. Currently you can use Java 11 as workaround.

**Comment by Drakmyth** [comment]
@CsCherrYY Is the fix being included in 3.8.1 for the issue mentioned in #1024 or for the Library Resolver issue my log is mentioning? Or are these in fact the same issue?

**Comment by CsCherrYY** [comment]
@Drakmyth It will be fixed in 3.8.1, but actually they are not the same issues. There will be a standalone fix for this bug. I'll comment on this issue once it's available.

**Comment by CsCherrYY** [comment]
@Drakmyth Could you help us to verify if the fix in #1031 is valid for your issue? You can follow 

1. Download https://github.com/CsCherrYY/vscode-test/blob/main/vscode-gradle-3.8.0.vsix
2. Uninstall vscode-gradle extension in your VSCode
3. Reload VSCode window
4. Manually install vscode-gradle via your downloaded vsix file, see
![extension](https://user-images.githubusercontent.com/45906942/137085423-39f058b8-d654-4a00-a4eb-443441176556.png)
5. To check if your error
```
Caused by: java.lang.NullPointerException: Cannot read the array length because "<local3>" is null
```
is still exists.

**Comment by Drakmyth** [comment]
I installed the v3.8.0 extension from the link you provided. The error no longer occurs and everything seems to be working normally!
